feat(curriculum): Added the coffee machine project

fhsinchy opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a [descriptive title](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request?id=prepare-a-good-pr-title) (**not** a vague title like `Update index.md`)

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Added the coffee machine project to teach OOP in the Python curriculum.

<!-- Feel free to add any additional description of changes below this line -->

Dario-DC commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,153 @@
+        }
+
+    def report(self):
+        print(f"Water: {self.resources['water']}ml")
+        print(f"Milk: {self.resources['milk']}ml")
+        print(f"Coffee: {self.resources['beans']}g")
```
```suggestion
        print(f"Water: {self.resources['water']} ml")
        print(f"Milk: {self.resources['milk']} ml")
        print(f"Coffee: {self.resources['beans']} g")
```
Adding some space to the output

Dario-DC commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,153 @@
+        self.profit = 0
+
+    def report(self):
+        print(f"Money: {self.CURRENCY}{self.profit}")
```
```suggestion
        print(f"Money: {self.CURRENCY} {self.profit}")
```

Dario-DC reviewed (commented):
Just few suggestions after a quick review

gikf commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,153 @@
+
+class MenuItem:
+
+    def __init__(self, name, water, milk, beans, cost):
+        self.name = name
+        self.cost = cost
+        self.ingredients = {
+            'water': water,
+            'milk': milk,
+            'beans': beans,
+        }
```
I think it would be better to pass ingredients as a list/dictionary, without specifically requiring all of them (ie. espresso already needs `0` water, so including it in ingredients is a bit obsolete).

This could be done in two ways:
```suggestion
    def __init__(self, name, ingredients, cost):
        self.name = name
        self.cost = cost
        self.ingredients = ingredients
```
Alternatively, with **kwargs:
```suggestion
    def __init__(self, name, cost, **ingredients):
        self.name = name
        self.cost = cost
        self.ingredients = ingredients
```

fhsinchy replied:
I'll need to think a bit more about that. I agree with this being outdated though.

gikf commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,153 @@
+            'milk': milk,
+            'beans': beans,
+        }
+
```
Some simple string representation?
```python
    def __str__(self):
        return self.name
```

fhsinchy replied:
This'd be very nice addition.

gikf commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,153 @@
+        ]
+
+    def get_all_menu_items(self):
+        options = ""
+        for item in self.menu:
+            options += f"{item.name}/"
+        return options
```
```suggestion
        return '/'.join(item.name for item in self.items)
```
Or, after adding `__str__` to `MenuItem`:
```suggestion
        return '/'.join(self.items)
```

- - -
Probably it also could be changed to `__str__` of the `Menu`.

gikf commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,153 @@
+    def make_payment(self, price):
+        money_received = float(input('How much money are you giving me? '))
+        if money_received >= price:
+            change = round(money_received - price, 2)
+            print(f"Here is {self.CURRENCY}{change} in change.")
+            self.profit += price
```
Just to point out, `change` is virtually infinite.

What about making machine that can accept only card payments, so no need to worry about change in the project?

fhsinchy replied:
Yeah why not but for now I think this will do. What do you say?

gikf replied:
It's not a big deal, it's just JS curriculum has certification project with cash register. Of course similarly complex implementation here is not needed at all.

There's a choice:
- Making simplified cash/change handling
- Making card only handling

To some extent both are simplifications, with spots where things are not like in the real thing. However I think with second option it wouldn't be as visible.

gikf commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,153 @@
+
+class Menu:
+
+    def __init__(self):
+        self.menu = [
+            MenuItem(name="latte", water=200, milk=150, beans=24, cost=2.5),
+            MenuItem(name="espresso", water=50, milk=0, beans=18, cost=1.5),
+            MenuItem(name="cappuccino", water=250, milk=50, beans=24, cost=3),
+        ]
```
More convenient would be creating menu items elsewhere:
```suggestion
    def __init__(self, items):
        self.items = items
```

If having default items is very needed, this pattern could be used to get the best from two worlds, but I'd advice against it.
```suggestion
    def __init__(self, items=None):
        if items is None:
            self.items = [
                MenuItem(name="latte", water=200, milk=150, beans=24, cost=2.5),
                MenuItem(name="espresso", water=50, milk=0, beans=18, cost=1.5),
                MenuItem(name="cappuccino", water=250, milk=50, beans=24, cost=3),
            ]
        else:
            self.items = items
```

fhsinchy replied:
What do you recommend in this situation?

gikf replied:
I think how it is now - creating menu items elsewhere - is the way to go. 👍

gikf commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,153 @@
+
+
+class Stash:
+    CURRENCY = '$'
```
Having this as class variable limits the further usage of the class. It could be changed to instance variable (possibly with default value), or just dropped.

Nevertheless currency and `cost` in `MenuItem` don't have any actual relation.

gikf commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,153 @@
+"""
+
+
+class Maker:
```
Maybe some methods to replenish the resources?

gikf commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,153 @@
+          WELCOME TO FCC COFFEE MACHINE!
+
+          ------ MENU ------
+          Espresso ($1.50)
+          Latte ($2.50)
+          Cappuccino ($3.00)
```
This is something that should be printed from actual items in the menu.

gikf commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
Ideally only the main part of code should print anything on it own, or based on the results/values from the class instance's methods. Methods instead of printing anything may return string.

gikf commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,153 @@
+from time import sleep
+
+"""
+The Maker class represents the coffee maker itself. It keep tracks of the water, milk,
+and beans in the machine. Makes coffee and reports on the resources.
```
Am I correct to assume there's the distinction between coffee maker and actual coffee machine?

gikf reviewed (commented):
There's couple things to look at to make usage of the classes easier. See specific comments below.

gikf commented:
Also how is this supposed to be used? Only as a script? It might be better to make it work both as a script and as a module (allow importing from elsewhere).

Dario-DC commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,153 @@
+    user_choice = str(
+        input(f'What would you like?\nOptions ({options}): ')).strip().lower()
+
+    if user_choice == 'off':
+        print('Shutting down!')
+        is_on = False
+    elif user_choice == 'report':
+        coffee_maker.report()
+        money_stash.report()
+    elif menu.find_menu_item(user_choice) is None:
+        print('\033[31mError. Please choose an available option.\033[m')
+    else:
+        beverage = menu.find_menu_item(user_choice)  # Encapsulates the result
+        sufficient_resources = coffee_maker.is_resource_sufficient(
+            beverage)  # TrueFalse result
+        sufficient_money = money_stash.make_payment(beverage.cost)
+        if sufficient_resources and sufficient_money:
+            print('Thank you! Allow us to make your beverage now...')
+            coffee_maker.make_coffee(beverage)
+            sleep(5)
```
Maybe this is not the best place to implement the structural pattern matching but, as discussed during the last meeting, it would be nice to show it instead of coding a menu that is very similar to what we have in the expense tracker.

```suggestion
    match user_choice:
        case 'off':
            print('Shutting down!')
            is_on = False
        case 'report':
            coffee_maker.report()
            money_stash.report()
        case user_choice if menu.find_menu_item(user_choice) is None:
            print('\033[31mError. Please choose an available option.\033[m')
        case _:
            beverage = menu.find_menu_item(user_choice)  # Encapsulates the result
            sufficient_resources = coffee_maker.is_resource_sufficient(beverage)  # TrueFalse result
            sufficient_money = money_stash.make_payment(beverage.cost)
            if sufficient_resources and sufficient_money:
                print('Thank you! Allow us to make your beverage now...')
                coffee_maker.make_coffee(beverage)
                sleep(5)
```

fhsinchy commented:
@gikf and @Dario-DC I have implemented some of the suggestions and updated the code. Please let me know what you think. I still have to move all the print statement to the main part of the code will do that tonight. Looking forward to your comments.

gikf commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,131 @@
+    print()
+
+    options = menu.get_all_menu_items()
+    user_choice = str(
+        input(f'What would you like?\nOptions ({options}): ')).strip().lower()
```
```suggestion
    user_choice = input(f'What would you like?\nOptions ({options}): ').strip().lower()
```
`str` is obsolete, `input` always returns string.

fhsinchy replied:
Yeah I forgot, my bad my bad.

gikf commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,131 @@
+        print(f'Money: ${self.profit}')
+
+    def make_payment(self, price: float):
+        money_received = float(input('How much money are you giving me? '))
```
This needs `try/except` to ensure number is entered.

fhsinchy replied:
I have added exception handling here as well as in the `find_menu_item` method. Previously in cases where the method returns `None` the print statement in the main section used to fail. Now the method throws an exception. I thought about doing something like creating a custom exception like "ItemNotFoundError" but went with a plain exception at the end. Please let me know your thoughts on that too.

Dario-DC commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,131 @@
+    def make_payment(self, price: float):
+        money_received = float(input('How much money are you giving me? '))
+        if money_received >= price:
+            change = round(money_received - price, 2)
```
Maybe instead of starting the program with `Stash(profit=0)`, the initial money in the machine could be something like `50`. Then we could have a maximum amount of money you can use to pay (instead of +∞), depending on if the money in the machine is greater than the change.

fhsinchy replied:
@gikf can you give me an idea of how the card based flow may work out please?

@Dario-DC yeah that can be done. Let me see.

gikf replied:
I think it could be something as simple as:
1. Display the price.
2. Ask to authorize card payment - ie. with `input`, value `'Y'` authorizes it, anything else cancels transaction.

fhsinchy commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,131 @@
+
+    def is_resource_sufficient(self, menu_item):
+        for ingredient, amount in menu_item.ingredients.items():
+            if amount >= self.resources.get(ingredient, 0):
```
If I put greater than, we'll get one less cup of coffee. I think this is fine.

gikf replied:
What do you mean? If required `amount` is greater than resource `ingredient`, coffee cannot be made. If they are equal, or resource is greater then coffee can be made.

fhsinchy commented on scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py:
```diff
@@ -0,0 +1,131 @@
+
+    def report(self):
+        for resource, amount in self.resources.items():
+            print(f'{resource}: {amount} ml')
```
Added a simple fix for this one. I could've done something more complex and appropriate but I think this may be enough for a project of this magnitude.

Dario-DC reviewed (commented):
A question about type annotations: are you also going to include return annotations?

fhsinchy commented:
> A question about type annotations: are you also going to include return annotations?

I didn't because Python can infer the type from return value itself.

gikf commented:
Python's default interpreter - CPython - do not care about annotations at all (as long as they don't break syntax). They can be incorrect, there won't be any complaints. Only IDEs or external tools might be trying to keep them correct, including inferring when that's possible.

Dario-DC commented:
> Python's default interpreter - CPython - do not care about annotations at all (as long as they don't break syntax). They can be incorrect, there won't be any complaints. Only IDEs or external tools might be trying to keep them correct, including inferring when that's possible.

Nevertheless we can check if they are correct with tests (if we want to introduce them). Of course, explaining that in a different environment this check should be done with some external tool.

fhsinchy commented:
I can remove them for now since we're using CPython on the platform.

Dario-DC reviewed (commented):
I am a bit confused about that:
![image](https://github.com/freeCodeCamp/CurriculumExpansion/assets/105294544/55176bbb-20e1-4d54-b475-1bcc98e7f573)

![image](https://github.com/freeCodeCamp/CurriculumExpansion/assets/105294544/f72e09ab-3b4e-414b-9705-f8b95bf5b588)

fhsinchy commented:
@Dario-DC I've implemented the changes you've suggested.

fhsinchy commented:
> I am a bit confused about that: ![image](https://private-user-images.githubusercontent.com/105294544/342267029-55176bbb-20e1-4d54-b475-1bcc98e7f573.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkyMjAwOTMsIm5iZiI6MTcxOTIxOTc5MywicGF0aCI6Ii8xMDUyOTQ1NDQvMzQyMjY3MDI5LTU1MTc2YmJiLTIwZTEtNGQ1NC1iNDc1LTFiY2M5OGU3ZjU3My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyNFQwOTAzMTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05YzhkYzQwNDc1OTliZjZmNWJhYjg1OTUzOTkxMDc3YzY1NWQyMmZjNGZjZGNmNzdjZWU5MzBhY2Y1NDlhMTU0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.DTdWwlk5U8aASorsoTCOig0Y_hmjc5cxFAdEzB_ERSY)
> 
> ![image](https://private-user-images.githubusercontent.com/105294544/342267275-f72e09ab-3b4e-414b-9705-f8b95bf5b588.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkyMjAwOTMsIm5iZiI6MTcxOTIxOTc5MywicGF0aCI6Ii8xMDUyOTQ1NDQvMzQyMjY3Mjc1LWY3MmUwOWFiLTNiNGUtNDE0Yi05NzA1LWY4Yjk1YmY1YjU4OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyNFQwOTAzMTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMDZkMWNkYTkxNTE5ZmE2ZjFhOWNiZGVmZWY0MzFjNjVkNjhmOWI4ODUyZjM2MzNmNmQ0ZjIxMjkyNWNiMjFhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.tgi43N_LUn3zIV_TxMxMM6fVmxcXDZhz3eFTUtYpxYE)

Not sure what's happening here. I've never come across this particular issue during my tests.

Dario-DC commented:
> Not sure what's happening here. I've never come across this particular issue during my tests.

It happened a couple of times and then it worked. Then it happened again and always with the same input :thinking:

fhsinchy commented:
> > Not sure what's happening here. I've never come across this particular issue during my tests.
> 
> It happened a couple of times and then it worked. Then it happened again and always with the same input 🤔

The amount of espresso I've ordered from the machine after your comment is enough to keep me awake for the rest of 2024 😄
